liftovervcf without sorting the variants

lindenb opened the pull request:
### Description

I'm trying to liftover gnomad/genome to hg38 but it takes to much time and memory. I'd rather have the resulting VCF file without sorting and then  possibly sort later with bcftools.

I added an option DISABLE_SORT to liftover vcf:  

 - now the `sorter` miight be null if DISABLE_SORT=true
 - I moved the variable `out` to a class member `accept`
 - I added a test that just write with/without DISABLE_SORT

now vcfliftover can write to stdout, variants are not sorted:

```
$ java -jar build/libs/picard.jar  LiftoverVcf INPUT=testdata/picard/vcf/LiftOver/testLiftoverBiallelicIndels.vcf OUTPUT=/dev/stdout CHAIN=testdata/picard/vcf/LiftOver/test.over.chain REJECT=/dev/null  REFERENCE_SEQUENCE=testdata/picard/vcf/LiftOver/dummy.reference.fasta DISABLE_SORT=true 2> /dev/null 
##fileformat=VCFv4.2
##FORMAT=<ID=GT,Number=1,Type=String,Description="Genotype">
##INFO=<ID=END,Number=1,Type=Integer,Description="Stop position of the interval">
##INFO=<ID=ReverseComplementedAlleles,Number=0,Type=Flag,Description="The REF and the ALT alleles have been reverse complemented in liftover since the mapping from the previous reference to the current one was on the negative strand.">
##INFO=<ID=SwappedAlleles,Number=0,Type=Flag,Description="The REF and the ALT alleles have been swapped in liftover due to changes in the reference. It is possible that not all INFO annotations reflect this swap, and in the genotypes, only the GT, PL, and AD fields have been modified. You should check the TAGS_TO_REVERSE parameter that was used during the LiftOver to be sure.">
##contig=<ID=chr1,length=540>
#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	Sample1
chr1	539	.	A	AAGGG	15676.17	PASS	END=539;ReverseComplementedAlleles	GT	0/0
chr1	421	.	CA	C	724.43	PASS	END=422;ReverseComplementedAlleles	GT	0/1
chr1	469	.	A	T	100	PASS	END=469;ReverseComplementedAlleles	GT	0/1
chr1	469	.	A	T	100	PASS	END=469;ReverseComplementedAlleles	GT	./.
chr1	421	.	CA	C	100	PASS	END=422;ReverseComplementedAlleles	GT	1/.
```

----

### Checklist (never delete this)

Never delete this, it is our record that procedure was followed. If you find that for whatever reason one of the checklist points doesn't apply to your PR, you can leave it unchecked but please add an explanation below.

#### Content
- [X] Added or modified tests to cover changes and any new functionality
- [X] Edited the README / documentation (if applicable)
- [X] All tests passing on Travis

#### Review
- [ ] Final thumbs-up from reviewer
- [ ] Rebase, squash and reword as applicable

For more detailed guidelines, see https://github.com/broadinstitute/picard/wiki/Guidelines-for-pull-requests

yfarjoun commented on src/main/java/picard/vcf/LiftoverVcf.java:
```diff
@@ -321,11 +327,23 @@ protected int doWork() {
+    		.setReferenceDictionary(walker.getSequenceDictionary())
+    		;
+        
+        if (DISABLE_SORT) {
```
can remove if statement by using 
`.modifyOption(Options.INDEX_ON_THE_FLY,!DISABLE_SORT).build()`

yfarjoun commented on src/test/java/picard/util/LiftoverVcfTest.java:
```diff
@@ -248,6 +248,44 @@ public void testWriteOriginalPosition(final boolean shouldWriteOriginalPosition)
+
+        //try to open with / without index
+        try (final VCFFileReader liftReader = new VCFFileReader(liftOutputFile, !disableSort)) {
+            // nothing
```
at least count the variants and make sure that you got the right number....

yfarjoun commented on src/main/java/picard/vcf/LiftoverVcf.java:
```diff
@@ -250,7 +253,10 @@
 
     private VariantContextWriter rejects;
+    /** the output VariantContextWriter */
+    private VariantContextWriter accept;
```
For consistency, can you rename this `accepts` (with `s`) or change this and `rejects` to `acceptedRecords` and `rejectedRecords` ?

yfarjoun commented on src/main/java/picard/vcf/LiftoverVcf.java:
```diff
@@ -321,11 +327,23 @@ protected int doWork() {
+            .modifyOption(Options.ALLOW_MISSING_FIELDS_IN_HEADER, ALLOW_MISSING_FIELDS_IN_HEADER)
+            .setOutputFile(OUTPUT)
+    		.setReferenceDictionary(walker.getSequenceDictionary())
+    		;
```
keep `;` in same line.

yfarjoun commented on src/main/java/picard/vcf/LiftoverVcf.java:
```diff
@@ -477,7 +508,11 @@ private void trackLiftedVariantContig(Map<String, Long> map, String contig) {
         trackLiftedVariantContig(liftedBySourceContig, source.getContig());
         trackLiftedVariantContig(liftedByDestContig, toAdd.getContig());
-        sorter.add(toAdd);
+        if (sorter != null) { //we're sorting the variants
```
```suggestion
        if (!DISABLE_SORT) { //we're sorting the variants
```

yfarjoun commented on src/main/java/picard/vcf/LiftoverVcf.java:
```diff
@@ -346,13 +364,21 @@ protected int doWork() {
-                TMP_DIR);
+        
+	    if (DISABLE_SORT) {
+	        log.info("Lifting variants over and writing the output file. Variants will not be sorted");
```
```suggestion
	        log.info("Lifting variants over and writing the output file. Variants will not be sorted.");
```

yfarjoun commented on src/main/java/picard/vcf/LiftoverVcf.java:
```diff
@@ -346,13 +364,21 @@ protected int doWork() {
+	        log.info("Lifting variants over and writing the output file. Variants will not be sorted");
+	    	
+	        sorter = null;
+	    	}
```
make one line :`} else {`

yfarjoun commented on src/test/java/picard/util/LiftoverVcfTest.java:
```diff
@@ -248,6 +248,44 @@ public void testWriteOriginalPosition(final boolean shouldWriteOriginalPosition)
+            // nothing
+            }
+        }
+    
```
too many NLs

yfarjoun commented on src/main/java/picard/vcf/LiftoverVcf.java:
```diff
@@ -441,21 +467,26 @@ protected int doWork() {
-        for (final VariantContext ctx : sorter) {
-            out.add(ctx);
-            progress.record(ctx.getContig(), ctx.getStart());
+        
```
extra nl

yfarjoun commented on src/main/java/picard/vcf/LiftoverVcf.java:
```diff
@@ -441,21 +467,26 @@ protected int doWork() {
+	        log.info("Writing out sorted records to final VCF.");
+	
+	        for (final VariantContext ctx : sorter) {
+	        	this.accept.add(ctx);
```
indentation is off.

yfarjoun commented on src/main/java/picard/vcf/LiftoverVcf.java:
```diff
@@ -477,7 +508,11 @@ private void trackLiftedVariantContig(Map<String, Long> map, String contig) {
         trackLiftedVariantContig(liftedByDestContig, toAdd.getContig());
-        sorter.add(toAdd);
+        if (sorter != null) { //we're sorting the variants
+        	sorter.add(toAdd);
```
please use spaces (4) not tabs for indent.

yfarjoun commented on src/test/java/picard/util/LiftoverVcfTest.java:
```diff
@@ -248,6 +248,44 @@ public void testWriteOriginalPosition(final boolean shouldWriteOriginalPosition)
+                "REJECT=" + rejectOutputFile.getAbsolutePath(),
+                "CHAIN=" + CHAIN_FILE,
+                "REFERENCE_SEQUENCE=" + REFERENCE_FILE,
+                "CREATE_INDEX=" + (!disableSort),
```
shouldn't the CREATE_INDEX be implicit? do you need to state it? if so, perhaps expand your provider to state indexing and sorting status?

lindenb commented:
@yfarjoun back to you.

Thank you for the review ,  I hope I removed all the formatting problems

> shouldn't the CREATE_INDEX be implicit? 

well one can imagine that you want a sorted VCF without index. I added a test 

```
        if (CREATE_INDEX && DISABLE_SORT) {
            log.error("CREATE_INDEX=true and DISABLE_SORT=true are mutually exclusive.");
            return 1;
         }
```

yfarjoun commented on src/main/java/picard/vcf/LiftoverVcf.java:
```diff
@@ -287,6 +287,11 @@ protected int doWork() {
         IOUtil.assertFileIsWritable(REJECT);
 
+        
+        if (CREATE_INDEX && DISABLE_SORT) {
```
``
this type of input validation should be in customCommandLineValidation()
``

lindenb replied:
in fact I realized that CREATED_INDEX is only for BAM-related tools, for LiftoverVCF, the index is always created event if CREATED_INDEX=false

yfarjoun commented:
what I meant is that your test only tests the explicit invocation, and not the implicit one...

yfarjoun commented on src/main/java/picard/vcf/LiftoverVcf.java:
```diff
@@ -181,6 +181,9 @@
     @Argument(doc = "INFO field annotations that should be deleted when swapping reference with variant alleles.", optional = true)
     public Collection<String> TAGS_TO_DROP = new ArrayList<>(LiftoverUtils.DEFAULT_TAGS_TO_DROP);
 
+    @Argument(doc = "Output VCF file will be written on the fly but it won't be sorted and indexed.", optional = true)
```
```suggestion
    @Argument(doc = "Output VCF file will be written on the fly but it won't be sorted nor indexed.", optional = true)
```

jmarshall replied:
“or” actually :smile:

yfarjoun replied:
I'm not native english speaker, nor do I care that much :-D

yfarjoun commented on src/main/java/picard/vcf/LiftoverVcf.java:
```diff
@@ -321,13 +332,16 @@ protected int doWork() {
-        out.writeHeader(outHeader);
+        this.acceptedRecords = new VariantContextWriterBuilder()
+            .modifyOption(Options.ALLOW_MISSING_FIELDS_IN_HEADER, ALLOW_MISSING_FIELDS_IN_HEADER)
+            .modifyOption(Options.INDEX_ON_THE_FLY,!DISABLE_SORT)
```
```suggestion
            .modifyOption(Options.INDEX_ON_THE_FLY, !DISABLE_SORT)
```

lindenb commented:
> what I meant is that your test only tests the explicit invocation, and not the implicit one...

Sorry, I'm not sure I understand :thinking: .  for example in a test like LiftoverVcfTest/testWriteOriginalPosition how is the condition implicit/explicit  tested ?

yfarjoun commented:
I guess I mean that you explicitly told the program to CREATE_INDEX rather than it doing whatever it thinks
